db_test.go

```diff
@@ -885,113 +885,105 @@ func (*mockCompactorFailing) Compact(dest string, dirs []string, open []*Block)
 		{MinTime: 300, MaxTime: 400},
 		{MinTime: 400, MaxTime: 500},
-		{MinTime: 500, MaxTime: 600},
+		{MinTime: 500, MaxTime: 600}, // Newest Block
 	}
 
-	var oldestBlockSize int64 = -1
-	var totalSize int64
 	for _, m := range blocks {
-		singleBlockSize := createPopulatedBlock(t, db.Dir(), 100, m.MinTime, m.MaxTime).Size()
-		if oldestBlockSize == -1 {
-			oldestBlockSize = singleBlockSize
-		}
-		totalSize += singleBlockSize
+		createPopulatedBlock(t, db.Dir(), 100, m.MinTime, m.MaxTime)
 	}
 
-	// Calculate the actual disk size using os.stat.
+	// Test that registered size matches the actual disk size.
+	testutil.Ok(t, db.reload())                       // Reload the db to register the new db size.
+	testutil.Equals(t, len(blocks), len(db.Blocks())) // Ensure all blocks are registered.
+	metrics := &dto.Metric{}                          // Use the the actual internal metrics.
+	testutil.Ok(t, db.metrics.blocksBytes.Write(metrics))
+	expSize := int64(metrics.Gauge.GetValue())
+	actSize := dbDiskSize(db.Dir())
+	testutil.Equals(t, expSize, actSize, "registered size doesn't match actual disk size")
+
+	// Decrease the max bytes limit so that a delete is triggered.
+	// Check total size, total count and check that the oldest block was deleted.
+	firstBlockSize := db.Blocks()[0].Size()
+	sizeLimit := actSize - firstBlockSize
+	db.opts.MaxBytes = sizeLimit // Set the new db size limit one byte smaller that the last block.
+	testutil.Ok(t, db.reload())  // Reload the db to register the new db size.
+	testutil.Ok(t, db.metrics.blocksBytes.Write(metrics))
+	testutil.Ok(t, db.metrics.sizeRetentionCount.Write(metrics))
+
+	expBlocks := blocks[1:]
+	actBlocks := db.Blocks()
+	expSize = int64(metrics.Gauge.GetValue())
+	actRetentCount := int(metrics.Counter.GetValue())
+	actSize = dbDiskSize(db.Dir())
+
+	testutil.Equals(t, 1, actRetentCount, "metric retention count mismatch")
+	testutil.Equals(t, actSize, expSize, "metric db size doesn't match actual disk size")
+	testutil.Assert(t, expSize <= sizeLimit, "actual size (%v) is expected to be less than or equal to limit (%v)", expSize, sizeLimit)
+	testutil.Equals(t, len(blocks)-1, len(actBlocks), "new block count should be decreased from:%v to:%v", len(blocks), len(blocks)-1)
+	testutil.Equals(t, expBlocks[0].MaxTime, actBlocks[0].meta.MaxTime, "maxT mismatch of the first block")
+	testutil.Equals(t, expBlocks[len(expBlocks)-1].MaxTime, actBlocks[len(actBlocks)-1].meta.MaxTime, "maxT mismatch of the last block")
+
+}
+
+func dbDiskSize(dir string) int64 {
 	var statSize int64
-	filepath.Walk(db.Dir(), func(path string, info os.FileInfo, err error) error {
-		// Directory sizes and meta files are not included in the size as of now.
-		if !info.IsDir() && !strings.HasSuffix(path, metaFilename) {
+	filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		// Include only index,tombstone and chunks.
+		if filepath.Dir(path) == chunkDir(filepath.Dir((filepath.Dir(path)))) ||
```

Comment:
Two sets of parentheses around `filepath.Dir(path)`?

Reply:
yes parent of the parent.

when we have /tmp/db/block1/chunk/0001 it becomes /tmp/db/block1 and than chunkDir() adds the chunk extension. so this ensures that the the file is indeed a chunk file. It is a little bit awkward, but this ensures that it will not brake if the dir name structure changes at some point.

Reply:
Sorry, I meant that it's `filepath.Dir((filepath.Dir(path)))` instead of `filepath.Dir(filepath.Dir(path))`. There's an extra pair of parentheses around it.

Reply:
aah yeah , just noticed that and updated.